components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AuthenticationHandler.java

```diff
@@ -101,10 +110,127 @@ protected void postAuthenticate(MessageContext messageContext, AuthenticationRes
 
                 if (user.getTenantDomain() != null && user.getTenantDomain().equalsIgnoreCase(PrivilegedCarbonContext
                         .getThreadLocalCarbonContext().getTenantDomain())) {
-                    PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(IdentityUtil.addDomainToName
-                            (user.getUserName(), user.getUserStoreDomain()));
+
+                    String domainAwareUserName = IdentityUtil.addDomainToName(user.getUserName(),
+                            user.getUserStoreDomain());
+                    String associatedUserRequestHeader = authenticationContext.getAuthenticationRequest().
+                            getHeader(ASSOCIATED_USER_ID_HEADER);
+                    if (StringUtils.isNotEmpty(associatedUserRequestHeader)) {
+                        User associatedUser = getAssociatedUser(authenticationResult, associatedUserRequestHeader);
+                        if (associatedUser != null && !isSameUser(user, associatedUser)) {
+                            setAssociatedUserInCarbonContext(user, associatedUser, authenticationResult);
+                            return;
+                        }
+                    }
+                    setUserInCarbonContext(domainAwareUserName);
                 }
             }
         }
     }
+
+    private User getAssociatedUser(AuthenticationResult authenticationResult, String associatedUserRequestHeader) {
+
+        User associatedUser = getUser(associatedUserRequestHeader);
+        if (associatedUser == null) {
+            log.error("Invalid user provided with the header: " + ASSOCIATED_USER_ID_HEADER);
+            setFailedAuthentication(authenticationResult);
+            return null;
+        }
+        return associatedUser;
+    }
+
+    private void setAssociatedUserInCarbonContext(User user, User associatedUser,
+                                                  AuthenticationResult authenticationResult) {
+
+        if (isUsersInSameTenant(user, associatedUser)) {
```

Comment:
better check if conditions and return rather than using else statement. That way the depth of the code will reduce and easy to navigate and understand the code.

Reply:
Fixed in b24e7a46ca66d94e33e0afd03cd4f787016d7325